spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/TwoStepsConversions.java

```diff
@@ -93,33 +98,61 @@ public TwoStepsConversions(CustomConversions customConversions, ObjectToKeyFacto
 
 		this.internalConversionService.addConverter(BYTE_ARRAY_TO_BLOB_CONVERTER);
 		this.internalConversionService.addConverter(BLOB_TO_BYTE_ARRAY_CONVERTER);
-
 	}
 
 	@Override
 	public <T> T convertOnRead(Object val, Class targetCollectionType,
 			Class targetComponentType) {
-		return convertOnRead(val, false, targetCollectionType, targetComponentType);
+		return (T) convertOnRead(val,
+				new EmbeddedStatus(0, EmbeddedStatus.EmbeddedType.NOT_EMBEDDED),
+				targetCollectionType,
+				ClassTypeInformation.from(targetComponentType));
 	}
 
 	@Override
-	public <T> T convertOnReadEmbedded(Object val, Class targetCollectionType,
-			Class targetComponentType) {
-		return convertOnRead(val, true, targetCollectionType, targetComponentType);
+	public <T> T convertOnRead(Object val, EmbeddedStatus embeddedStatus,
+			TypeInformation targetTypeInformation) {
+		TypeInformation componentTypeInformation;
+		if (targetTypeInformation.isCollectionLike()) {
+			componentTypeInformation = targetTypeInformation.getComponentType();
+		}
+		else if (embeddedStatus
+				.getEmbeddedType() == EmbeddedStatus.EmbeddedType.EMBEDDED_MAP) {
+			componentTypeInformation = (TypeInformation) targetTypeInformation
+					.getTypeArguments().get(1);
+		}
+		else {
+			componentTypeInformation = targetTypeInformation;
+		}
+		return convertOnRead(val, embeddedStatus,
+				targetTypeInformation.isCollectionLike() ? targetTypeInformation.getType()
+						: null,
+				componentTypeInformation);
 	}
 
-	@SuppressWarnings("unchecked")
-	private <T> T convertOnRead(Object val, boolean isEmbedded,
-			Class targetCollectionType, Class targetComponentType) {
+	private <T> T convertOnRead(Object val, EmbeddedStatus embeddedStatus,
+			Class targetCollectionType, TypeInformation targetComponentType) {
 		if (val == null) {
 			return null;
 		}
-		BiFunction<Object, Class<?>, T> readConverter = isEmbedded
-				? this::convertOnReadSingleEmbedded
-				: this::convertOnReadSingle;
+		BiFunction<Object, TypeInformation<?>, ?> readConverter;
+		switch (embeddedStatus.getEmbeddedType()) {
+		case EMBEDDED_MAP:
+			readConverter = (x, y) -> convertOnReadSingleEmbeddedMap(x, y,
+					embeddedStatus.getEmbeddedMapDepthAllowance());
+			break;
+		case SINGULAR_EMBEDDED:
```

Comment:
the singular and collection embedded actually don't have different use cases so they can just be one enum value.